### PR TITLE
Fix property enumeration order

### PR DESCRIPTION
**Pull request by ruben-ayrapetyan**
Implementation of `ecma_op_object_get_property_names` interface that iterates an object's properties in the following order (https://github.com/Samsung/jerryscript/issues/287#issuecomment-118139791):
 - integer-named properties in index ascending order
 - string-named properties in creation order

Improvement of performance of item 3.l of arrays' `[[DefineOwnProperty]]` (`ecma_op_array_object_define_own_property`), using the new interface.

The changes fix the following test262 tests:
 - ch15/15.2/15.2.3/15.2.3.4/15.2.3.4-4-1
 - ch15/15.2/15.2.3/15.2.3.4/15.2.3.4-4-2
 - ch15/15.2/15.2.3/15.2.3.4/15.2.3.4-4-44
 - ch15/15.2/15.2.3/15.2.3.5/15.2.3.5-4-39
 - ch15/15.2/15.2.3/15.2.3.5/15.2.3.5-4-40
 - ch15/15.2/15.2.3/15.2.3.7/15.2.3.7-6-a-93-1
 - ch15/15.2/15.2.3/15.2.3.7/15.2.3.7-6-a-93-2
 - ch15/15.2/15.2.3/15.2.3.7/15.2.3.7-6-a-93-3
 - ch15/15.2/15.2.3/15.2.3.7/15.2.3.7-6-a-93-4
 - ch15/15.2/15.2.3/15.2.3.14/15.2.3.14-2-8
 - ch15/15.2/15.2.3/15.2.3.14/15.2.3.14-3-1
 - ch15/15.2/15.2.3/15.2.3.14/15.2.3.14-3-3
 - ch15/15.2/15.2.3/15.2.3.14/15.2.3.14-4-1
 - ch15/15.2/15.2.3/15.2.3.14/15.2.3.14-5-13
 - ch15/15.2/15.2.3/15.2.3.14/15.2.3.14-5-b-1
 - ch15/15.4/15.4.5/15.4.5.2/S15.4.5.2_A3_T4

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp [hunk 1]
```diff
@@ -1222,63 +1222,45 @@ ecma_builtin_array_prototype_object_sort (ecma_value_t this_arg, /**< this argum
...
-      continue;
-    }
+    ecma_string_t *property_name_p = ecma_get_string_from_value (*iter.current_value_p);
 
     uint32_t index;
-    if (ecma_string_get_array_index (property_name_p, &index) && index < len)
+    bool is_index = ecma_string_get_array_index (property_name_p, &index);
+    JERRY_ASSERT (is_index);
+
+    if (index < len)
```
Comment (egavrin) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.cpp]: Please invert this `if` and eliminate 1271L `else`

**Comment by egavrin** [comment]
Good to me

**Comment by zherczeg** [comment]
Am I see right that you reverse properties during enumeration? Clever.

LGTM

**Comment by ruben-ayrapetyan** [comment]
Merged (da47c671e71bfa3a6da42b7b012579e220b2f080)

**Comment by idkiller** [comment]
I modified my patch with using `ecma_op_object_get_property_names`.

**Pull request closed without merging**